### DISCUSSION TITLE
Contents of "cmake-out"

### DISCUSSION THREAD

**Discussion by epasveer** [googleapis/google-cloud-cpp#14051] [Q&A] [answered]
After cloning and building the C++ sdk, I see the build artifacts (libraries)  appear in "cmake-out" directory.  However, the include files are not there.

Am I wrong to expect the include files (like google-cloud-cpp/cmake-out/) to not be there?

Here are my build commands after cloning the repo.
```
https://github.com/googleapis/google-cloud-cpp

Building with CMake
This library requires CMake >= 3.10. If you are planning to install the libraries please consult the [packaging guide](https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md), these instructions will NOT produce artifacts that you can put in /usr/local, or share with your colleagues.

From the top-level directory of google-cloud-cpp run these commands:

git -C $HOME clone https://github.com/microsoft/vcpkg.git
env VCPKG_ROOT=$HOME/vcpkg $HOME/vcpkg/bootstrap-vcpkg.sh
cmake -S . -B cmake-out/ -DCMAKE_TOOLCHAIN_FILE=$HOME/vcpkg/scripts/buildsystems/vcpkg.cmake
cmake --build cmake-out -- -j $(nproc)

The binary artifacts, such as examples, will be placed in cmake-out/.
```

**Comment by coryan** [comment, marked as answer]
First: if you are using vcpkg.io to get the dependencies of `google-cloud-cpp`, then you may want to use it to install `google-cloud-cpp` too. Mixing manually compiled libraries and libraries compiled with a package manager is full of pitfalls.

If you tell us more about what are you trying to do, maybe we can point you to the right documentation for your use-case.  For example, if you wanted to compile `google-cloud-cpp` from source then /doc/packaging.md is probably what you want. If you want to use vcpkg.io, then just the instructions from that package manager should work. You would just do something like `vcpkg install google-cloud-cpp` and use what it produces.

----

> However, the include files are not there.
> Am I wrong to expect the include files (like google-cloud-cpp/cmake-out/) to not be there?

I think you mean "Am i wrong to expect the include files to **be** there?".  

In any case, maybe this can help:

* `cmake --build` just compiles the binary artifacts (and other generated files, e.g. protobuf-generated `*.pb.h` files)
* `cmake --install` compiles the binary artifacts, headers, and support files to a destination

If you want to install the libraries and headers you want something like:

```shell
cmake -S . -B cmake-out --toolchain $HOME/vcpkg/scripts/buildsystems/vcpkg.cmake
cmake --build cmake-out -- -j $(nproc)
cmake --install cmake-out --prefix /usr/local
```

However, that will leave things in /usr/local that depend on the `vcpkg` binaries in `$HOME/vcpkg`.  In almost every case that is a bad idea.

**Reply by epasveer** [reply]
Thanks for the quick answer.

I think I'm just missing this step.
```
cmake --install cmake-out --prefix /usr/local
```

For now, we don't want the artifacts to end up in /usr/local -- so I'll put it somewhere in my home directory.  I'll play around with this command to see what I get.

**Reply by coryan** [reply]
> For now, we don't want the artifacts to end up in /usr/local -- so I'll put it somewhere in my home directory.

Whatever works for you is fine by me, of course.  I feel obliged to mention that `$HOME/vcpkg/vcpkg install google-cloud-cpp` will leave the libraries and headers in `$HOME/vcpkg/installed/...`.  That may be a quicker way to play around with the libraries.

Likewise, you may want to browse: 

https://cloud.google.com/cpp/docs/setup#cmake-with-conda

or

https://cloud.google.com/cpp/docs/setup#cmake-with-vcpkg

Conda downloads binary packages, you can be up and running in one minute (that was a speed run, but you get the idea: no time waiting for the compiler).

**Reply by epasveer** [reply]
That command worked perfectly!  Thanks.

**Reply by epasveer** [reply]
Thanks for the added info.  We're just getting into gcloud's C++ api.  I'm sure we'll revisit the building of the C++ sdk in the future.  I've marked this dicsussion.
